homeassistant/components/binary_sensor/mqtt.py

```diff
@@ -162,8 +193,17 @@ def state_message_received(_topic, payload, _qos):
 
             self.async_schedule_update_ha_state()
 
-        await mqtt.async_subscribe(
-            self.hass, self._state_topic, state_message_received, self._qos)
+        if self._state_topic != old_state_topic:
+            if self._state_unsub is not None:
+                self._state_unsub()
+            self._state_unsub = await mqtt.async_subscribe(
+                self.hass, self._state_topic, state_message_received,
+                self._qos)
+
+    async def async_will_remove_from_hass(self):
+        """Unsubscribe when removed."""
+        if self._state_unsub is not None:
```

Comment:
I wonder if we can make a helper for this.

Something like:

```python
# constructor
self.sub_state = None

# Somewhere later
self.sub_state = subscribe_topics(self.sub_state, {
  'state': { 'topic': state_topic, 'qos': XX, 'handler': self._handle_state_updated},
})
# This method will check for each key if any topics are different and does the unsub and re-sub dance if necessary.
# return value is an opaque object that only the subscribe/unsubscribe methods know about.

# Also can be used for unsub
unsub_topics(self.sub_state)
```

Reply:
Let's be nice for future readers of our code and do `mqtt/subscription.py`

Reply:
OK, implemented now.